### PR TITLE
fix: send an empty-string pagination cursor verbatim on every list adapter

### LINKED ISSUE #2220: Four */list adapters drop an empty-string cursor, silently re-requesting page one
Raised in the **v2.5.0 milestone-merge review** (#2215) and confirmed against the source. Filed here rather than fixed in the merge PR, whose tree is byte-identical to `origin/v2/main`.

## The asymmetry

`core/mcp/inspectorClient.ts` builds the five `*/list` request params two different ways:

| Adapter | Guard | `cursor: ""` |
| --- | --- | --- |
| `listTools` | `cursor !== undefined` | **sent** |
| `listPrompts` | `cursor ? { cursor } : {}` | dropped |
| `listResources` | `cursor ? { cursor } : {}` | dropped |
| `listResourceTemplates` | `cursor ? { cursor } : {}` | dropped |
| `listRequestorTasks` | `cursor ? { cursor } : {}` | dropped |

A truthiness check cannot distinguish "no cursor" from "the cursor is the empty string". MCP cursors are **opaque strings** — the spec places no constraint on their content, so `""` is a valid `nextCursor` a server may legitimately return, and the client is required to send it back verbatim.

## Consequence

When a server paginates with an empty-string cursor, the four truthiness adapters drop it and re-request **page one**. Depending on the caller that is either a list that silently stops at the first page, or a loop that fetches page one forever because `nextCursor` never advances. Nothing surfaces an error — the request is well-formed, it just asks the wrong question.

`listTools` is already correct, and its inline comment states the reasoning: dropping the cursor asks for page one again.

## This is already suspected in-tree

`clients/web/src/lib/protocolReplay.ts` encodes the asymmetry as a known quirk rather than a rule, and says as much:

```
// And an **empty** one survives on `tools/list` alone. `listTools` builds
// its params with `cursor !== undefined`, carrying `""` deliberately —
// its own comment explains that dropping it asks for page one again. The
// other four adapters use a truthiness check and drop it. That asymmetry
// looks like a latent bug in those four rather than an intention, but
// this function's job is to describe what the dispatch *does*, so it
// reports the empty cursor as dropped where it would be dropped.
```

That comment made the right call for `protocolReplay` — describing the dispatch faithfully is its job — but it also records an unfiled bug. This is the issue it should have pointed at.

## Fix

Bring the four adapters to `listTools`'s guard:

```ts
...(cursor !== undefined ? { cursor } : {}),
```

Then simplify `getReplayableParams` in `protocolReplay.ts`: the `method === "tools/list" || params.cursor !== ""` special case exists only to mirror the asymmetry, and once the adapters agree the condition collapses to `typeof params.cursor === "string"`. Update the long comment there to match rather than leaving it describing a state that no longer exists.

Cover it with a test server fixture that paginates via an empty-string cursor, driven through at least one of the four — the request params are the assertion, so a unit test on the adapter is enough, but the fixture is what keeps it honest end to end.

Reported by Copilot on #2215; the finding appeared in the review summary without an accompanying inline comment, and was verified from the source before filing.

### PR DESCRIPTION
**Pull request by cliffhall**
Closes #2220

## The bug

An MCP cursor is an **opaque string** — the spec constrains neither its content nor its length — so `""` is a `nextCursor` a server may legitimately return, and the client is required to send it back verbatim.

`core/mcp/inspectorClient.ts` built the list request params two different ways. `listTools` used `cursor !== undefined` (and so does the later `listSkills`); `listPrompts`, `listResources`, `listResourceTemplates` and `listRequestorTasks` used `cursor ? { cursor } : {}`, a truthiness check that cannot distinguish "no cursor" from "the cursor is the empty string".

Against a server that paginates with an empty-string cursor, those four dropped it and re-requested **page one** — either a list that silently stops at the first page, or a walk that fetches page one forever because `nextCursor` never advances. Nothing surfaced an error: the request was well-formed, it just asked the wrong question.

The asymmetry was already suspected in-tree — `clients/web/src/lib/protocolReplay.ts` encoded it as a known quirk and said in a comment that it looked like a latent bug. This is the issue it should have pointed at.

## The change

- **`core/mcp/inspectorClient.ts`** — the four adapters now build params with `cursor !== undefined`, matching `listTools`. Each carries the reason inline.
- **`clients/web/src/lib/protocolReplay.ts`** — `replayableParams`'s `method === "tools/list" || params.cursor !== ""` special case existed only to mirror the asymmetry. With the adapters agreed it collapses to `typeof params.cursor === "string"`, and the long comment is rewritten rather than left describing a state that no longer exists.

## Coverage

- **`clients/web/src/test/core/mcp/inspectorClient-list-cursor.test.ts`** — the outbound params of all five adapters: `""` carried verbatim, an absent cursor sending no `cursor` key at all, a non-empty cursor forwarded unchanged. `listTools` is included as the control, so a regression in the guard it has always had goes red here too.
- **`test-servers/configs/empty-cursor-http.json`** — a new fixture (`emptyStringCursor`, threaded through `load-config` → `resolve-config` → a shared cursor codec in the four pagination handlers) that hands out `""` as the cursor for page two and a numeric index for page three. No other fixture produces that shape; every other one's cursor is a non-empty index string, which the buggy guard happened to carry correctly. Off by default, so the existing `pagination-http.json` keeps its numeric cursors.
- **`clients/web/src/test/integration/mcp/empty-cursor.test.ts`** — walks all three of that server's lists end to end, asserting the server really hands back `""`, that sending it back yields the *second* four items rather than the first four again, and that the walk still terminates.
- **`docs/test-servers.md`** — fixture row plus a repro section.

**Mutation check** — reverting the fix on one adapter fails exactly its two cases and nothing else:

| Guard removed from | Result |
| --- | --- |
| `listPrompts` | 2 failed (`listPrompts sends an empty-string cursor verbatim`, `prompts/list advances past an empty-string cursor`), 17 passed |

No UI surface changed — the fix is in the shared protocol layer, and the web/TUI list panels render whatever the adapters return — so there are no screenshots. `npm run local:gate` passes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01EN2UfrkQZvtrviMtqURRm4

### CODE REVIEW

File: core/mcp/inspectorClient.ts [hunk 1]
```diff
@@ -3045,7 +3045,13 @@ export class InspectorClient extends InspectorClientEventTarget {
       throw new Error("Client is not connected");
     }
     const result = await this.client.request(
-      { method: "tasks/list", params: cursor ? { cursor } : {} },
+      {
+        method: "tasks/list",
+        // `!== undefined`, not truthiness: a cursor is opaque and `""` is a
+        // legal value a server may hand back. Dropping it asks for page one
+        // again, so a caller walking pages would loop on the first page.
+        params: cursor !== undefined ? { cursor } : {},
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply core/mcp/inspectorClient.ts:3053]: Fixed — good catch, and it was the only remaining walk on the truthiness side.

`ManagedRequestorTasksState.refresh()` drives the legacy `tasks/list` pagination itself, so a `""` `nextCursor` ended the loop before the corrected adapter was ever reached a second time. Both of its cursor tests are now `=== undefined`:

- `while (cursor)` → `while (cursor !== undefined)`
- `this.tasks = cursor ? [...this.tasks, ...page] : page` → `cursor === undefined ? page : [...this.tasks, ...page]`. That second one is the same bug one step later: had the walk continued, page two would have *replaced* the accumulated list rather than extended it.

I audited the rest of the stores rather than fixing only the one you named; every other walk already distinguished `undefined` (`pagedToolsState`, `pagedPromptsState`, `pagedResourcesState`, `pagedResourceTemplatesState`, `pagedRequestorTasksState`, `managedSkillsState`), so this was the last one.

New case in `managedRequestorTasksState.test.ts`, paginating `"" → "c2" → end`. It asserts **which cursor each call received** (`[undefined, "", "c2"]`), not just the task count — a walk that terminated correctly but re-requested page one would otherwise still show three tasks and pass. `FakeInspectorClient.listRequestorTasks` gained a declared signature so `mock.calls` carries that argument.

Mutation check — reverting either guard on its own fails exactly this case:

| Guard reverted | Result |
| --- | --- |
| `while (cursor)` | 1 failed, 28 passed |
| `cursor ? […] : page` | 1 failed, 28 passed |

3b4eb447. `npm run local:gate` passes.

**Comment by cliffhall** [comment]
## Review round 1 — resolved

Mirroring the inline reply here, since inline threads go hidden once the fix is pushed.

**Copilot (1 comment, `core/mcp/inspectorClient.ts`):** the legacy task refresh still treated cursors as booleans — `managedRequestorTasksState.ts` used `cursor ? …` and `while (cursor)` — so a `""` `nextCursor` ended the walk before the corrected adapter was reached, leaving the task list on page one.

**Done** (`3b4eb447`). Both guards are now `=== undefined`. The second one is the same bug one step later: had the walk continued, page two would have *replaced* the accumulated list rather than extended it.

I audited the rest of the stores rather than fixing only the one named — every other walk already distinguished `undefined` (`pagedToolsState`, `pagedPromptsState`, `pagedResourcesState`, `pagedResourceTemplatesState`, `pagedRequestorTasksState`, `managedSkillsState`), so this was the last one on the truthiness side.

Covered by a new case in `managedRequestorTasksState.test.ts` paginating `"" → "c2" → end`, asserting **which cursor each call received** (`[undefined, "", "c2"]`) rather than just the task count — a walk that terminated correctly but re-requested page one would otherwise still show three tasks and pass. `FakeInspectorClient.listRequestorTasks` gained a declared signature so `mock.calls` carries that argument.

| Guard reverted | Result |
| --- | --- |
| `while (cursor)` | 1 failed, 28 passed |
| `cursor ? […] : page` | 1 failed, 28 passed |

Rounds 2 and 3 both came back 🟢 with 0 comments. `npm run local:gate` passes.
